runtime/storage/storage.go

```diff
@@ -0,0 +1,151 @@
+	"gocloud.dev/blob"
+	"gocloud.dev/blob/gcsblob"
+	"gocloud.dev/gcp"
+)
+
+type Client struct {
+	dataDirPath  string
+	bucketConfig *gcsBucketConfig
+	prefixes     []string
+}
+
+func New(dataDir string, bucketCfg map[string]any) (*Client, error) {
+	c := &Client{
+		dataDirPath: dataDir,
+	}
+
+	if len(bucketCfg) != 0 {
+		gcsBucketConfig := &gcsBucketConfig{}
+		err := mapstructure.WeakDecode(bucketCfg, gcsBucketConfig)
+		if err != nil {
+			return nil, err
+		}
+		c.bucketConfig = gcsBucketConfig
+	}
+	return c, nil
+}
+
+func MustNew(dataDir string, bucketCfg map[string]any) *Client {
+	c, err := New(dataDir, bucketCfg)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func (c *Client) WithPrefix(prefix ...string) *Client {
+	newClient := &Client{
+		dataDirPath:  c.dataDirPath,
+		bucketConfig: c.bucketConfig,
+	}
+	newClient.prefixes = append(newClient.prefixes, c.prefixes...)
+	newClient.prefixes = append(newClient.prefixes, prefix...)
+	return newClient
+}
+
+func (c *Client) DataDir(elem ...string) string {
+	paths := []string{c.dataDirPath}
+	if c.prefixes != nil {
+		paths = append(paths, c.prefixes...)
+	}
+	paths = append(paths, elem...)
+	return filepath.Join(paths...)
+}
+
+func (c *Client) TempDir(elem ...string) string {
+	paths := []string{c.dataDirPath}
+	if c.prefixes != nil {
+		paths = append(paths, c.prefixes...)
+	}
+	paths = append(paths, "tmp")
+	paths = append(paths, elem...)
+	return filepath.Join(paths...)
+}
+
+func (c *Client) OpenBucket(ctx context.Context, elem ...string) (*blob.Bucket, bool, error) {
+	if c.bucketConfig == nil {
+		return nil, false, nil
+	}
+	// Init dataBucket
+	client, err := c.newGCPClient(ctx)
+	if err != nil {
+		return nil, false, fmt.Errorf("could not create GCP client: %w", err)
+	}
+
+	bucket, err := gcsblob.OpenBucket(ctx, client, c.bucketConfig.Bucket, nil)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to open bucket %q: %w", c.bucketConfig.Bucket, err)
+	}
+	var prefix string
+	for _, p := range c.prefixes {
+		prefix = prefix + p + "/"
+	}
+	for _, e := range elem {
+		prefix = prefix + e + "/"
+	}
+	if prefix == "" {
+		return bucket, true, nil
+	}
+	return blob.PrefixedBucket(bucket, prefix), true, nil
+}
+
+func AddInstance(c *Client, instanceID string) error {
+	if c.prefixes != nil {
+		return fmt.Errorf("storage: should not call AddInstance with prefixed client")
+	}
+
+	c = c.WithPrefix(instanceID)
+	err := os.Mkdir(c.DataDir(), os.ModePerm)
+	if err != nil && !errors.Is(err, fs.ErrExist) {
+		return fmt.Errorf("could not create instance directory: %w", err)
+	}
+
+	// recreate instance's tmp directory
+	tmpDir := c.TempDir()
+	if err := os.RemoveAll(tmpDir); err != nil {
+		return fmt.Errorf("could not remove instance tmp directory: %w", err)
+	}
+	if err := os.Mkdir(tmpDir, os.ModePerm); err != nil && !errors.Is(err, fs.ErrExist) {
+		return err
+	}
+
+	return nil
+}
+
+func RemoveInstance(c *Client, instanceID string) error {
```

Comment:
> Actually for 1. The temp dir is part of every instance so we can't delete directory in New

I think iterating and deleting is okay (not great). An alternative is generating temp paths in `{root}/tmp/{prefix}` so we can easily clear temp data by deleting `tmp`? 

Or another idea is to get a temp root by calling `os.TempDir` and using `{temp root}/{prefix}` for temp paths? Then the OS will automatically clean it for us on restarts. What do you think?

> Not a strong opinion but it felt simpler to just return paths and let callers take care of creating/deleting directories if any sub directory is required similar to the current API.

Yeah that also sounds good to me, but then I don't think we need `AddInstance` right? I mean if the caller anyways is required to call `MkDirAll` before using a path, then `AddInstance` is not needed. 

But I also think it will be fine just to create the directory (but yes then it needs to return the error), callers can check contents if they really need to know if they are the first or not.

> Yes cleanup of instance data in bucket needs to be added. But do we want in a sync path or via bg job similar to DeleteAssets ?

Ideally a background job, at some point we will move all of `DeleteInstance` to be invoked idempotently by a background job. But until then, it's fine to do it sync :)

Reply:
> An alternative is generating temp paths in {root}/tmp/{prefix} so we can easily clear temp data by deleting tmp?

This will mean that `tmp` data is stored outside of instance directory in`data_dir`. At some point we wanted to store all data including any tmp files in a single directory. 

> Or another idea is to get a temp root by calling os.TempDir and using {temp root}/{prefix} for temp paths? Then the OS will automatically clean it for us on restarts. What do you think?

This also has the problem specified above. We may also want to use mounted disks for storing tmp files like file generated by bigquery connector, exported files etc.

> I mean if the caller anyways is required to call MkDirAll before using a path, then AddInstance is not needed.

I think the idea was that base directory for each instance is always created but if any subdirectory is still required then caller need to create that. I think this is similar to how go's API like `TempDir` and `UserCacheDir` operates ?

Reply:
I think the main problem in current `storage` abstraction is that the concept of `instance` is neither coupled nor decoupled and that is what `AddInstance` and `RemoveInstance` tries to compensate but I can go ahead with suggested changes as well.

Reply:
> But I also think it will be fine just to create the directory (but yes then it needs to return the error), callers can check contents if they really need to know if they are the first or not.

Given the elem is a variable list, another consideration is whether we return error if last element is a clash or any element clashes.

Reply:
> This will mean that tmp data is stored outside of instance directory indata_dir. At some point we wanted to store all data including any tmp files in a single directory.

The main thing is having some way to monitor how much temp data is being written by an instance, but it doesn't strictly speaking need to be in the same directory as the persistent data (but yes that's how we did it initially). As long as all temp data goes in one directory with sub-directories for each instance, then we will still be able to add stats/monitoring of per-instance temp data in the future.

> This also has the problem specified above. We may also want to use mounted disks for storing tmp files like file generated by bigquery connector, exported files etc.

I think this has been more of a weakness of our previous approach. It would actually be better to use an ephemeral local SSD for temp data than a networked persistent disk. I think this would be a good change to make, but we should give infra a heads up so they can ensure that `os.TempDir` is an ephemeral SSD (it probably  already is) and has decent capacity.

> I think the idea was that base directory for each instance is always created but if any subdirectory is still required then caller need to create that.

We almost never access the `data` directory directly, we are anyway appending the connector name to it, so the directory doesn't exist even now I think?

> I think this is similar to how go's API like TempDir and UserCacheDir operates ?

The docs for `TempDir` say "The directory is neither guaranteed to exist nor have accessible permissions." Looking at the implementation, the same seems to be the case for `UserCacheDir`.

> Given the elem is a variable list, another consideration is whether we return error if last element is a clash or any element clashes.

If you decide to create the directory in these calls, then I think the normal behavior/errors of `MkdirAll` would be expected.

Reply:
Sure. Using `os.TempDir` sounds good in that case.